libs/@guardian/identity-auth/src/identityAuth.ts

```diff
@@ -0,0 +1,146 @@
+ *
+ * This provides minimal functionality for the Guardian's OAuth2.0 implementation.
+ *
+ * It provides the following functionality when instantiated:
+ * - A `tokenManager` for managing oauth tokens
+ * - A `token` for requesting new oauth tokens
+ * - An `authStateManager` for managing the current authentication state
+ *
+ * @param {IdentityAuthOptions} options - The options for IdentityAuth
+ */
+export class IdentityAuth<
+	AC extends CustomClaims = CustomClaims,
+	IC extends CustomClaims = CustomClaims,
+> {
+	#options: IdentityAuthOptions;
+	#oauthUrls: OAuthUrls;
+	#emitter: Emitter;
+
+	public tokenManager: TokenManager<AC, IC>;
+	public token: Token<AC, IC>;
+	public authStateManager: AuthStateManager<AC, IC>;
+
+	constructor(options: IdentityAuthOptions) {
+		this.#options = options;
+		this.#oauthUrls = {
+			authorizeUrl: `${this.#options.issuer}/v1/authorize`,
+			tokenUrl: `${this.#options.issuer}/v1/token`,
+			keysUrl: `${this.#options.issuer}/v1/keys`,
+		};
+
+		this.#emitter = new Emitter();
+		this.token = new Token<AC, IC>(this.#options, this.#oauthUrls);
+		this.tokenManager = new TokenManager<AC, IC>(this.#emitter, this.token);
+		this.authStateManager = new AuthStateManager<AC, IC>(
+			this.#emitter,
+			this.tokenManager,
+		);
+	}
+
+	/**
+	 * @name isSignedInWithAuthState
+	 * @description Checks if the user is signed in, and updates the auth state as necessary, returns the current auth state
+	 *
+	 * This performs side effects.
+	 *
+	 * 1. If the user has a GU_SO cookie, they have recently signed out, so we should clear their tokens (side effect)
+	 * 2. If user tokens already exist, they are signed in
+	 * 3. If the user doesn't have tokens, but they have a GU_U cookie, they are "maybe" signed in
+	 *   a. We can try to get tokens without prompting/redirecting the user for credentials (side effect)
+	 * 4. If the user doesn't have tokens or a GU_U cookie, they are not signed in
+	 *
+	 * @returns `AuthState` - Returns the current authentication state
+	 */
+	public async isSignedInWithAuthState(): Promise<IdentityAuthState<AC, IC>> {
+		try {
+			// if the user has a GU_SO cookie, they have recently signed out, so we should clear their tokens
+			// the GU_SO cookie will be automatically cleared when the user signs back in
+			if (getCookie({ name: 'GU_SO', shouldMemoize: true })) {
+				this.tokenManager.clear();
+				return {
+					isAuthenticated: false,
+				};
+			}
+
+			// if user tokens already exist, they are signed in
+			const authState = this.authStateManager.getAuthState();
+			if (
+				authState?.isAuthenticated &&
+				authState.accessToken &&
+				authState.idToken
+			) {
+				// validate the id token and access token to make sure auth state is still valid
+				await this.token.verifyToken(authState.idToken, authState.accessToken);
+
+				// if the id token is valid, return the auth state
+				return authState;
```

Comment:
Doesn't this line always return the `authState`, irrespective of whether `this.token.verifyToken` says that the ID token is valid?

Reply:
Oh, `verifyToken` throws, I see!